test/e2e/common.go

```diff
@@ -0,0 +1,370 @@
+	workapiv1 "github.com/open-cluster-management/api/work/v1"
+	"github.com/open-cluster-management/registration-operator/pkg/helpers"
+	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1beta1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	eventuallyTimeout            = 60 // seconds
+	eventuallyInterval           = 1  // seconds
+	clusterManagerNamespace      = helpers.ClusterManagerNamespace
+	klusterletDefaultNamespace   = helpers.KlusterletDefaultNamespace
+	bootstrapHubKubeConfigSecret = helpers.BootstrapHubKubeConfigSecret
+	hubRegistrationDeployment    = "cluster-manager-registration-controller"
+	hubWebhookDeployment         = "cluster-manager-registration-webhook"
+	operatorNamespace            = "open-cluster-management"
+	klusterletOperator           = "klusterlet"
+)
+
+var (
+	kubeClient         kubernetes.Interface
+	clusterCfg         *rest.Config
+	operatorClient     operatorclient.Interface
+	clusterClient      clusterclient.Interface
+	workClient         workv1client.Interface
+	bootstrapHubSecret *corev1.Secret
+)
+
+func createKlusterlet(name, clusterName, agentNamespace string) (realClusterName string) {
+	var klusterlet = &operatorapiv1.Klusterlet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: operatorapiv1.KlusterletSpec{
+			RegistrationImagePullSpec: "quay.io/open-cluster-management/registration",
+			WorkImagePullSpec:         "quay.io/open-cluster-management/work",
+			ExternalServerURLs: []operatorapiv1.ServerURL{
+				{
+					URL: "https://localhost",
+				},
+			},
+			ClusterName: clusterName,
+			Namespace:   agentNamespace,
+		},
+	}
+
+	Expect(name).NotTo(BeEmpty())
+
+	// create agentNamespace and bootstrap-hub-kubeconfig secret
+	if agentNamespace != "" && agentNamespace != klusterletDefaultNamespace {
+		_, err := kubeClient.CoreV1().Namespaces().Get(context.TODO(), agentNamespace, metav1.GetOptions{})
+		if err != nil && errors.IsNotFound(err) {
+			namespace := v1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: agentNamespace,
+				},
+			}
+			_, err := kubeClient.CoreV1().Namespaces().Create(context.TODO(), &namespace, metav1.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		}
+
+		secret := bootstrapHubSecret.DeepCopy()
+		secret.ObjectMeta.ResourceVersion = ""
+		secret.SetNamespace(agentNamespace)
+		_, err = kubeClient.CoreV1().Secrets(agentNamespace).Create(context.TODO(), secret, metav1.CreateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	// create klusterlet CR
+	realKlusterlet, err := operatorClient.OperatorV1().Klusterlets().Create(context.TODO(), klusterlet, metav1.CreateOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	realClusterName = realKlusterlet.Spec.ClusterName
```

Comment:
this is probably where your method should end, and the 'does the ManagedCluster exist?' concern factored into a separate method.

Reply:
I split the old method into CreateKlusterlet, GetCreatedManagedCluster, ApproveCSR, AcceptsClient, CheckManagedClusterStatus.